### ISSUE TITLE
iOS: dark mode

### ISSUE THREAD

**Issue by Dejal** [samuelclay/NewsBlur#825]
Every app that someone might read in bed needs a dark mode.  We should add this (I know of many people who won't use an app without such a feature, like CGP Grey).

All views should be updated to support the option of a black background with light text, in addition to the current colors.

For stories especially, it'd be good to support four background colors, like the Safari view controller: white, sepia, gray & black.

It'd be good to have a preference to automatically switch based on the brightness of the screen, in addition to a preference for the background color.

A two-finger vertical swipe to manually switch would also be a nice feature (like in Tweetbot etc).

**Comment by samuelclay** [comment]
Yes, I like the Tweetbot gesture. An auto-color depending on brightness is a fine addition, although I hope it isn't too hard to figure out brightness and compensate for it. This should have an option for auto-enabling dark mode based on ambient brightness, as well as an option for *which* theme to use for either auto-enabling or with a gesture. We should also have the gesture able to be turned off.

**Comment by Dejal** [comment]
Okie dokie artichokie.  After epic changes, I'm now pretty much finished with this.

There are new Theme preferences, to set the current theme, whether or not to switch automatically based on the screen brightness, whether or not to switch via two-finger swipe up (to light) or down (to dark), like a light switch.  And which themes to use for each.

![screen shot 2015-12-07 at 15 46 08](https://cloud.githubusercontent.com/assets/1318849/11643582/b8eb62bc-9cf9-11e5-90dd-702964f2d221.png)

(If you're reading the GitHub notification email, check out the web version to see the images.)

There are two light themes, named "Light" (the existing theme) and "Sepia" (with a yellowed tint), and two dark themes, named "Medium" (gray) and "Dark" (black).

Here's a sample of the dark theme:

![simulator screen shot 2015-12-07 15 53 52](https://cloud.githubusercontent.com/assets/1318849/11643778/0e19a8e2-9cfb-11e5-9362-530131dcfcdf.png)

And here's the story view, with the menu popover displayed; notice the color selection circles at the bottom of the popover:

![simulator screen shot 2015-12-07 15 52 37](https://cloud.githubusercontent.com/assets/1318849/11643760/ee3abf3e-9cfa-11e5-9ff2-4b4483736df3.png)

Finally, here's sepia:

![simulator screen shot 2015-12-07 15 58 20](https://cloud.githubusercontent.com/assets/1318849/11643830/62f6a04a-9cfb-11e5-9c8e-ccfb4fc4c181.png)

The one remaining thing to do is finish styling the story CSS.  I thought that you might like to do that, since the colors are largely a matter of preference, but if you prefer I'd be happy to have a go at it.  I did some basic styling of each theme, but there's more to do.

On the implementation side, there is a new `ThemeManager` class that manages the themes.  The old `UIColorFromRGB()` macro is still used, but now calls a `ThemeManager` method to adjust the color as needed.  There are now also `UIColorFromFixedRGB()` and `UIColorFromLightDarkRGB()` macros, too. 

For the stories, in addition to the `storyDetailView.css` CSS file, there are now `storyDetailViewDark.css`, `storyDetailViewMedium.css` and `storyDetailViewSepia.css` override files, which only include color stuff, and overrides the base story CSS file.  If you want to tackle styling those themes, those are the files to edit.

Take a look, and let me know what you think.  I hope you like it!  It's possible I missed a view, or some use case.  I'll leave this issue open for now.

**Comment by samuelclay** [comment]
We should fix how stories load in on an empty canvas. about:blank won't cut it, as the screen flashes white before the black story loads in.

**Comment by samuelclay** [comment]
Excellent, I'm going to try and work through the colors next week. 5.0.3 was just submitted and should be approved soon.

**Comment by nriley** [comment]
Some stuff I've noticed on iPad (aside from color selection/contrast issues, which @samuelclay mentions and obviously will need a lot of tweaking):

- On the main screen the unread/focused numbers disappear when switching between themes
- Open popovers don't update to reflect the theme
- The popover arrow is always white/light gray, which looks bizarre
- The back button arrowhead is black rather than light gray after switching (opacity?)
- The iOS status bar text doesn't switch to white for the dark theme
- The bars under the icon portion of menus shouldn't be there in the dark theme; they look weird

**Comment by Dejal** [comment]
Thanks for this.  I'll see what I can do to improve those.

**Comment by nriley** [comment]
Retested and updated the comment to show status.

**Comment by Dejal** [comment]
Thanks.  

- I can't reproduce the unread etc numbers disappearing.  They correctly change color.  Can you provide screenshots and/or clarification?

- Yep, easier and perfectly reasonable to close popovers when changing themes, especially since you're swiping in a view behind them.

- Yeah, the custom popovers don't change their arrows; I need to continue replacing them with the modern style.

**Comment by nriley** [comment]
> I can't reproduce the unread etc numbers disappearing. They correctly change color. Can you provide screenshots and/or clarification?

![simulator screen shot dec 14 2015 5 56 22 pm](https://cloud.githubusercontent.com/assets/68303/11796718/c96b4d22-a28b-11e5-8356-a050982d2a57.png)

See how there's no numbers visible?  This happens after two-finger dragging to change to the dark theme.

**Comment by Dejal** [comment]
Ah, gotcha.  I had thought you meant the table view counts, not the navigation bar header counts.  Fixed.

**Comment by nriley** [comment]
Nice, looks like everything functional is fixed!  Now for all that poor contrast, ugly reverse shadows and gradients, non-matching icons, ... :-)

**Comment by samuelclay** [comment]
If you want to give it a shot, I'm happy to take any PRs. Four themes is quite a lot and I'm struggling to find time to work on it. I hope to get to it in the new year, since this won't be launching anytime before that.

There's a lot of places that the text shadows, gradients, and borders need to be reconsidered.

**Comment by Dejal** [comment]
I'm happy to help with any of it.  Just let me know.  If you like, I could do a first pass on the story CSS, then get more feedback on other things that need tweaking.

**Comment by nriley** [comment]
Seems we're getting closer here! I fixed a couple things this morning, but some remaining stuff:

The font popover still has bright lines in the icon section like used to affect the menu popover:

![bright lines in font popover](https://cloud.githubusercontent.com/assets/68303/13028473/141ed244-d23e-11e5-8b5f-3525117eaba0.png)

Wonder if it would be possible to coalesce some of the drawing on theme changes on the feed pages; there's quite a lot of flashing involved (even in the simulator).

Several issues remain with shadow drawing.

**Comment by nriley** [comment]
Font popover is fixed.

It seems coalescing drawing between the Web page and surrounding frame is hard, or I'm insufficiently competent at figuring out how.  I tried clearing the story so you get a theme-colored background prior to the drawing happening in the new theme, but that doesn't look much better.  Even on the simulator, redrawing the page takes a long time and is flickery.

Some of the icon shadows and the medium theme could potentially use more work but this seems like it's close enough to being usable.  Perhaps time for a public beta of 5.1?

**Comment by samuelclay** [comment]
I need to fix the icon shadows (by removing them) and fix the share dialog and notifier (bottom bar on feed list).

**Comment by nriley** [comment]
The notifier should be fixed by 289f484539dfca30bd9045f14eb4ad623a3e05e6.  I never use the share dialog but definitely it needs work too :-)

**Comment by nriley** [comment]
Things are looking much better! One glitch I noticed was that the handle on the splitter bar on iPad seems to get rather wide when switching between certain themes:
![wide handle](https://cloud.githubusercontent.com/assets/68303/13548174/e859d476-e2b7-11e5-89e7-0588cc7b1f6a.PNG)

**Comment by samuelclay** [comment]
Speaking of iPad, that view also has the accidentally pushed down story traversal buttons. I can open a new ticket for that, but I'd really like to fix that as well.

**Comment by Dejal** [comment]
I'll take a look tomorrow.

**Comment by Dejal** [comment]
Fixed the size bar issue.  Still looking into the traverse button issue; it's a curly one.

**Comment by Dejal** [comment]
Traverse buttons fixed.  No more bouncing up & down, or positioning too low.  Tricky to find, but obvious once I did.  (Sorry for the delay in getting back to this; I've been busy with support from my [Time Out 2.0](http://www.dejal.com/timeout/) release.)

**Comment by samuelclay** [comment]
I thought it would be obvious, but I couldn't find it either.

**Comment by nriley** [comment]
OK, trying out the iPad version for a change.  Some more stuff:

- The bar that reads "NEWSBLUR" at the top right doesn't update with theme changes.
- When you tap on the gear or + icons on the main screen, they revert to the theme color as it was at startup and if you don't actually change themes, it stays that way.
- The folder arrowheads don't theme properly on startup, only after changing themes.

And something I'm not sure is theme-related but is definitely broken:
- If I tap a folder to go to it on startup, nothing happens.  (Sites seem OK.)

**Comment by nriley** [comment]
And a fun rotation drawing glitch. Aiee. Sorry!

![img_0670](https://cloud.githubusercontent.com/assets/68303/13658325/0891cd86-e644-11e5-9bde-eea0ddbb69f8.PNG)

**Comment by Dejal** [comment]
I've just fixed all of those, other than the rotation issue.

**Comment by Dejal** [comment]
The rotation issue just needed a judicious refresh of the table.

**Comment by nriley** [comment]
Thanks! I've not noted any new issues since then; I fixed a couple of popover drawing issues today that have been bugging me for ages. How about a beta now? :-)

**Comment by samuelclay** [comment]
Last bug I've seen is with the folder icon sizes in the Add Site popover. That popover also has some size issues when adding a new subfolder (being hidden by the autocomplete menu).

**Comment by nriley** [comment]
Ah yes, I never use that either… I prototyped a bit of trying to use MenuViewController to display the folder list here as the old UI never really looked right after iOS 7. Doesn't quite work yet but here's at least a screenshot...
![add site](https://cloud.githubusercontent.com/assets/68303/13745102/916b18b6-e9c2-11e5-9275-145b196e6777.png)

**Comment by Dejal** [comment]
Want me to look into the Add Site popover?  I agree that using the new folder list would be huge improvement.

**Comment by samuelclay** [comment]
Yep, I agree that the new folder list would be a better interface.

**Comment by Dejal** [comment]
Righto.  @nriley would you be okay with me redoing this?  I don't want to stomp on your changes if you've done more than a quick prototype.

**Comment by nriley** [comment]
No problem, it was just 20 min worth of work to see how hard it would be. I need to be doing other things anyway :-)

**Comment by Dejal** [comment]
Done.  Hope you like it.  So, ready for a beta now?  It'd be great to get this update out!

**Comment by samuelclay** [comment]
I'm thinking about releasing directly to the App Store. This is a big release and I want to see if I can give a blog an exclusive. Anyway, what new features am I missing for the release notes?

* Three new themes: dark, grey, and sepia
* New organizer
* New mute feeds dialog
* New folder and feed chooser (useful for future features)
* New contributors: David & Nicholas
* New icons
* New menus for deleting and muting feeds

**Comment by Dejal** [comment]
- Custom domain support
- Mark all as read (long press etc)
- Search improvements
- Share extension (don't forget this, in a separate branch)

**Comment by samuelclay** [comment]
Wait, the share extension is done? I thought we were waiting on 5.2 for it. Also, what are the search improvements?

**Comment by nriley** [comment]
- Keyboard navigation support (#826) although I do need to do another pass because a few more areas, now including startup, that have since broken the first responder since I last checked (which cause keyboard equivalents to go dead). I will go through the app and fix this over the weekend unless either of you want to do it earlier!

**Comment by samuelclay** [comment]
Go for it, there's no rush. I've got an Android release coming a bit sooner.

**Comment by Dejal** [comment]
The share extension was done ages ago (early in the 5.1 work).  Could use more testing, though (and missing the icon; probably just an into.plist tweak).

Actually, I think the search improvements were in 5.0 (changes to filtering feeds for free vs paid users).

**Comment by nriley** [comment]
Fixed the first responder for popovers, which seems to handle the vast majority of issues. I'll test a bit more over the next couple days with my iPad and keyboard to see if I can catch anything else.

A couple more theme related issues: the add site "+" icon still has an unsightly looking shadow in the gray theme. Its popover arrow also isn't being themed - banged my head against this one for a while but no luck figuring it out; hopefully @Dejal will have better luck!

**Comment by Dejal** [comment]
@samuelclay will need to update the '+' icon image.

The popover arrow was a curly one.  It was only resetting the color when the keyboard (or keyboard shortcut bar) was shown, which is apparently an iOS 8 issue.  The workaround was to move the pop origin a bit to the right.

**Comment by nriley** [comment]
Fixed some popover-related stuff and now nothing requires using the escape key (which doesn't exist on many iPad keyboards including Apple's Smart Keyboard); you can use ⌘. instead, which at least matches the Mac.

And yet more theming bugs…

The popover arrow fix seems to work on iPad but not iPhone. If you're testing in the simulator, it can help to to turn off the hardware keyboard so the issue gets triggered.

When launching in a non-default theme then switching back to the white theme on iPad, the separators between story table rows draw dark rather than light:
![dark rows](https://cloud.githubusercontent.com/assets/68303/13899876/05989fdc-edcf-11e5-86a5-7486a70c722b.png)

**Comment by Dejal** [comment]
I've fixed the iPhone popover issue, but I can't reproduce the iPad separator issue.

**Comment by nriley** [comment]
Okay, here are some systematic instructions. I’ve tested this both on the simulator and the device.

1. Switch to either the gray or black theme.
2. Force quit NewsBlur.
3. Relaunch NewsBlur.
4. Switch to the white or yellow theme.

You should now note that the separator lines are dark as you see above.

I've not found any more keyboard bugs this weekend, so I think we're in good shape.

**Comment by samuelclay** [comment]
Is iPad Pro support finished as well? What did we have to do to support the iPad pro?

**Comment by samuelclay** [comment]
I've noticed that the above bug also effects the separator between the story list and the story detail on iPad pro simulator. I booted in dark theme and switched to light theme.

<img width="1024" alt="screen shot 2016-03-20 at mar 20 8 02 37 pm" src="https://cloud.githubusercontent.com/assets/44229/13909821/d26d5536-eed6-11e5-9f89-5c933a1d3559.png">

**Comment by nriley** [comment]
Yes the keyboard support is ideal for an iPad with a keyboard case, so you can have a similar-to-Web reading environment. As long as you use the in-app browser you should be able to navigate exclusively that way (for my reading style at least).

I've never tested it on an actual iPad Pro because I don't own one (and don't plan on getting one, at least the big variety; may replace my iPad Air with whatever comes out today).

**Comment by Dejal** [comment]
Sorry for the delay replying.  I think the biggest change for iPad Pro support was fixing the signup/login view; see #788.  The keyboard support is definitely useful, too.

I do have an iPad Pro with keyboard case, and a couple of iPad minis; each have their place.  I really like the Pro for couch usage.  And I like the extra screen space of the big Pro; the smaller one announced today doesn't really appeal to me; I have the minis and iPhone for more mobile situations.

I'll look into the theme switching issues probably tomorrow.  (Do let me know if you're ever in a hurry; I can adjust priorities as needed.)

**Comment by samuelclay** [comment]
I'm almost never in a hurry, so no worries on timing. The only rush I'm ever in is for fixing recently shipped releases, which I've had to consistently do ever since v1.0. (Who hasn't?)

Today during a few demos I did notice the theme bugs when switching themes. It felt like the gesture worked more reliably than the menu color chooser. But that's only anecdotal, I don't remember the exact repro. Should be easy to spot though. It was mainly when switching themes right after starting the app, something that happens in regular use.
